src/backend/distributed/commands/common.c

```diff
@@ -0,0 +1,284 @@
+ * changed locally.
+ *
+ * We perform the same precondition checks as before to skip this operation if any of the
+ * failed during preprocessing. Since we already raised an error on other checks we don't
+ * have to repeat them here, as they will never fail during postprocessing.
+ *
+ * When objects get altered they can start depending on undistributed objects. Now that
+ * the objects has been changed locally we can find these new dependencies and make sure
+ * they get created on the workers before we send the command list to the workers.
+ */
+List *
+PostprocessAlterDistributedObjectStmt(Node *stmt, const char *queryString)
+{
+	const DistributeObjectOps *ops = GetDistributeObjectOps(stmt);
+	Assert(ops != NULL);
+
+	ObjectAddress address = GetObjectAddressFromParseTree(stmt, false);
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	if (ops->featureFlag && *ops->featureFlag == false)
+	{
+		/* not propagating when a configured feature flag is turned off by the user */
+		return NIL;
+	}
+
+	EnsureDependenciesExistOnAllNodes(&address);
+
+	return NIL;
+}
+
+
+/*
+ * PreprocessDropDistributedObjectStmt is a general purpose hook that can propagate any
+ * DROP statement.
+ *
+ * DROP statements are one of the few DDL statements that can work on many different
+ * objects at once. Instead of resolving just one ObjectAddress and check it is
+ * distributed we will need to lookup many different object addresses. Only if an object
+ * was _not_ distributed we will need to remove it from the list of objects before we
+ * recreate the sql statement.
+ *
+ * Given that we actually _do_ need to drop them locally we can't simply remove them from
+ * the object list. Instead we create a new list where we only add distributed objects to.
+ * Before we recreate the sql statement we put this list on the drop statement, so that
+ * the SQL created will only contain the objects that are actually distributed in the
+ * cluster. After we have the SQL we restore the old list so that all objects get deleted
+ * locally.
+ *
+ * The reason we need to go through all this effort is taht we can't resolve the object
+ * addresses anymore after the objects have been removed locally. Meaning during the
+ * postprocessing we cannot understand which objects were distributed to begin with.
+ */
+List *
+PreprocessDropDistributedObjectStmt(Node *node, const char *queryString,
+									ProcessUtilityContext processUtilityContext)
+{
+	DropStmt *stmt = castNode(DropStmt, node);
+
+	/*
+	 * We swap the list of objects to remove during deparse so we need a reference back to
+	 * the old list to put back
+	 */
+	List *originalObjects = stmt->objects;
+
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	QualifyTreeNode(node);
+
+	List *distributedObjects = NIL;
+	List *distributedObjectAddresses = NIL;
+	Node *object = NULL;
+	foreach_ptr(object, stmt->objects)
+	{
+		/* TODO understand if the lock should be sth else */
```

Comment:
We need at least AccessShareLock to prevent the object from being dropped concurrently. 

I think theoretically, we are a bit more likely to deadlock against concurrent transactions that first do SELECT and then DDL. 

If we did AccessExclusive lock here, we would prevent the SELECT from happening concurrently. 

On the other hand, there are always some deadlock possibilities when it comes to DDL, e.g. DROP TABLE locks tables in one order, and a concurrent SELECT locks them in another order.